File: leetcode/130-surrounded-regions/main.py
"""
    1st approach: recursive dfs

    Time    O(2n)
    Space   O(n) hashtable
    416 ms, faster than 7.72%
"""


class Solution(object):
    def solve(self, grid):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        # seen 2d array
        seen = []
        for i in range(len(grid)):
            seen.append(len(grid[0])*[False])
        # for each island, find out if they should be captured
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if seen[i][j] == False and grid[i][j] == 'O':
                    touchBound = self.dfs(grid, i, j, seen)
                    if touchBound == False:
                        self.markCapture(grid, i, j)

    def dfs(self, grid, i, j, seen):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]):
            return True
        if seen[i][j] == True:
            return False
        seen[i][j] = True
        if grid[i][j] == 'O':
            up = self.dfs(grid, i-1, j, seen)
            down = self.dfs(grid, i+1, j, seen)
            left = self.dfs(grid, i, j-1, seen)
            right = self.dfs(grid, i, j+1, seen)
            return up or down or left or right
        return False

    def markCapture(self, grid, i, j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == 'X':
            return
        if grid[i][j] == 'O':
            grid[i][j] = 'X'
            self.markCapture(grid, i-1, j)
            self.markCapture(grid, i+1, j)
            self.markCapture(grid, i, j-1)
            self.markCapture(grid, i, j+1)


a = [
    ["X", "X", "X", "X"],
    ["X", "O", "O", "X"],
    ["X", "X", "O", "X"],
    ["X", "O", "X", "X"],
]
Solution().solve(a)
print(a)

a = [
    ["X", "X", "X", "O"],
    ["X", "O", "O", "X"],
    ["X", "X", "O", "X"],
    ["X", "O", "X", "X"],
]
Solution().solve(a)
print(a)

a = [
    ["X", "X", "X", "X", "X"],
    ["X", "O", "O", "O", "X"],
    ["X", "X", "X", "O", "X"],
    ["X", "O", "X", "O", "O"],
    ["X", "X", "X", "X", "X"],
]
Solution().solve(a)
print(a)

File: leetcode/130-surrounded-regions/main.go
package main

/*
	1st approach: recursive dfs

	Time    O(2n)
	Space   O(n) hashtable
	56 ms, faster than 7.94%
*/
func solve(grid [][]byte) {
	// visited
	visited := make([][]bool, len(grid))
	for i := range visited {
		visited[i] = make([]bool, len(grid[i]))
	}
	// traverse island
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == 'O' && !visited[i][j] {
				tocuhBound := dfs(grid, i, j, visited)
				if tocuhBound == false {
					markCapture(grid, i, j)
				}
			}
		}
	}
}

func dfs(grid [][]byte, i int, j int, visited [][]bool) bool {
	if i < 0 || i+1 > len(grid) || j < 0 || j+1 > len(grid[0]) {
		return true
	}
	if visited[i][j] == true {
		return false
	}
	visited[i][j] = true
	if grid[i][j] == 'O' {
		up := dfs(grid, i-1, j, visited)
		down := dfs(grid, i+1, j, visited)
		left := dfs(grid, i, j-1, visited)
		right := dfs(grid, i, j+1, visited)
		return up || down || left || right
	}
	return false
}

func markCapture(grid [][]byte, i int, j int) {
	if i < 0 || i+1 > len(grid) || j < 0 || j+1 > len(grid[0]) {
		return
	}
	if grid[i][j] == 'O' {
		grid[i][j] = 'X'
		markCapture(grid, i-1, j)
		markCapture(grid, i+1, j)
		markCapture(grid, i, j-1)
		markCapture(grid, i, j+1)
	}
}

func main() {

}
